tilecloud_chain/__init__.py

```diff
@@ -1069,6 +1133,51 @@ def __call__(self, tile):
         return tile
 
 
+class DatabaseLogger:  # pragma: no cover
+
+    def __init__(self, connection, schema, table, layer):
+        self.connection = connection
+        self.full_table = '{}.{}'.format(schema, table)
+        self.layer = layer
+        self.run = None
+
+    def __call__(self, tile):
+        if self.run is None:
+            run = (
+                '(SELECT COALESCE(MAX(run), 0) + 1 FROM {} '
+                'WHERE layer = %(layer)s)'
+            ).format(self.full_table)
+        else:
+            run = self.run
+
+        if tile.error:
+            action = 'error'
+        else:
+            action = 'create'
+
+        with self.connection.cursor() as cursor:
+            try:
+                cursor.execute(
+                    'INSERT INTO {}(layer, run, action, tile)'
+                    'VALUES (%(layer)s, {}, %(action)s::varchar(7), %(tile)s)'
+                    'RETURNING run'.format(self.full_table, run),
+                    {'layer': self.layer, 'action': action, 'tile': str(tile.tilecoord)}
+                )
+                self.run, = cursor.fetchone()
+            except psycopg2.IntegrityError:
+                self.connection.rollback()
+                if action != 'create':
```

Comment:
As I understood the requirements, a regeneration would have a different run number, so it wouldn't be able to find the log for the previous attempt to update it.

This `if` is because with the way the logger is called, it will only ever receive (for a given tile) a series of successes, sometimes followed by an error. If we would update and the action is 'create', it already is 'create' in the database